internal/pkg/gateway/registry.go

```diff
@@ -93,41 +90,113 @@ func (reg *registry) endorsers(channel string, chaincode string) ([]*endorser, e
+					continue
+				}
+
+				groupPeers = append(groupPeers, &endorserState{peer: peer, endorser: endorser, height: height})
+			}
+
+			// If the number of available endorsers less than the quantity required, try the next layout
+			if len(groupPeers) < int(quantity) {
+				abandonLayout = true
+				break
 			}
+
 			// sort by decreasing height
 			sort.Slice(groupPeers, sorter(groupPeers, reg.localEndorser.address))
 
-			peerGroup := groupPeers[0:quantity]
-			r = append(r, peerGroup...)
+			receivers = append(receivers, groupPeers[0:quantity]...)
+		}
+
+		if abandonLayout {
+			// try the next layout
+			continue
 		}
+
 		// sort the group of groups by decreasing height (since Evaluate will just choose the first one)
-		sort.Slice(r, sorter(r, reg.localEndorser.address))
-
-		for _, peer := range r {
-			if peer.endpoint == reg.localEndorser.address {
-				endorsers = append(endorsers, reg.localEndorser)
-			} else if endorser, ok := reg.remoteEndorsers[peer.endpoint]; ok {
-				endorsers = append(endorsers, endorser)
-			} else {
-				reg.logger.Warnf("Failed to find endorser at %s", peer.endpoint)
+		sort.Slice(receivers, sorter(receivers, reg.localEndorser.address))
+
+		for _, peer := range receivers {
+			endorsers = append(endorsers, peer.endorser)
+		}
+		return endorsers, nil
+	}
+
+	return nil, fmt.Errorf("failed to select a set of endorsers that satisfy the endorsement policy")
+}
+
+// endorsersForOrgs returns a set of endorsers owned by the given orgs for the given chaincode on a channel.
+func (reg *registry) endorsersForOrgs(channel string, chaincode string, endorsingOrgs []string) ([]*endorser, error) {
+	err := reg.registerChannel(channel)
+	if err != nil {
+		return nil, err
+	}
+
+	endorsersByOrg := make(map[string][]*endorserState)
+
+	members := reg.discovery.PeersOfChannel(common.ChannelID(channel))
+	for _, member := range members {
+		endpoint := member.Endpoint
+		// find the endorser in the registry for this endpoint
+		var endorser *endorser
+		if endpoint == reg.localEndorser.address {
+			endorser = reg.localEndorser
+		} else if e, ok := reg.remoteEndorsers[endpoint]; ok {
+			endorser = e
+		} else {
+			reg.logger.Warnf("Failed to find endorser at %s", endpoint)
+			continue
+		}
+		// check the org membership
+		if contains(endorsingOrgs, endorser.mspid) && member.Properties != nil {
+			for _, installedChaincode := range member.Properties.Chaincodes {
+				// only consider the peers that have our chaincode installed
+				if installedChaincode.Name == chaincode {
+					endorsersByOrg[endorser.mspid] = append(endorsersByOrg[endorser.mspid], &endorserState{endorser: endorser, height: member.Properties.LedgerHeight})
+				}
 			}
 		}
 	}
 
+	var endorsers []*endorser
+	for _, es := range endorsersByOrg {
+		// sort by decreasing height and select the highest one from each org
+		sort.Slice(es, sorter(es, reg.localEndorser.address))
+		if len(es) > 0 {
+			endorsers = append(endorsers, es[0].endorser)
+		}
```

Comment:
Isn't it a noop?

Reply:
If, for any reason, the endorser returned by the discovery service is not in the gateway registry, then this _could_ be an empty array.  Perhaps if a new endorser had just come online and not been added to the registry yet.  There's still a TODO comment in `registerChannels()` to handle this type of dynamic update.

Reply:
Discussed offline and @andrew-coleman is going to push a new patch. Holding it from merging for now.

Reply:
Thanks Manish, I've added an error check that will report which orgs fail to find endorsing peers, and added extra unit tests